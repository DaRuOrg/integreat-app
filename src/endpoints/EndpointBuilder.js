import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'
import StateMapperBuilder from './StateMapperBuilder'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _url
  _mapper
  _stateMapperBuilder
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name
    this._stateMapperBuilder = new StateMapperBuilder(this)

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a url to the builder
   * @param url The url
   * @return {EndpointBuilder} The builder itself
   */
  withUrl (url) {
    this._url = url
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds a state mapper builder to this builder
   * @return {StateMapperBuilder} The state mapper of this builder
   */
  withStateMapper () {
    return this._stateMapperBuilder
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder}  The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._stateMapperBuilder) {
      throw Error('You have to set a state mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._url, this._mapper, this._stateMapperBuilder.build(), this._refetchLogic)
  }
}

export default EndpointBuilder
