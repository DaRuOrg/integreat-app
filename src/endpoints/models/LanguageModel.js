class LanguageModel {
  constructor (code, name) {
    this._name = code
    this._code = name
  }

  get code () {
    return this._name
  }

  get name () {
    return this._code
  }
}

export default LanguageModel
