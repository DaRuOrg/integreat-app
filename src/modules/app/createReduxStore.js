// @flow

import type { Store } from 'redux'
import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import { createLogger } from 'redux-logger'

import { languageReducer, uiDirectionReducer } from '../i18n/reducers'
import toggleDarkModeReducer from '../theme/reducers'
import Payload from '../endpoint/Payload'

export type ActionType<T> = { type: string, payload: Payload<T> }

// todo: Change type to correct State type,
// https://blog.callstack.io/type-checking-react-and-redux-thunk-with-flow-part-2-206ce5f6e705
const createReduxStore = (initialState: {} = {}): Store<any, any> => {
  /**
   * The middlewares of this app, add additional middlewares here
   */
  const middlewares = []

  if (__DEV__) {
    middlewares.push(createLogger()) // Logs all state changes in console
  }

  const rootReducer = combineReducers({
    uiDirection: uiDirectionReducer,
    language: languageReducer,
    darkMode: toggleDarkModeReducer
  })

  const enhancers = compose(applyMiddleware(...middlewares))

  return createStore(rootReducer, initialState, enhancers)
}

export default createReduxStore
