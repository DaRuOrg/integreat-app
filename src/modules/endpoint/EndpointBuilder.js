import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _routerToUrlMapper
  _mapper
  _responseOverride
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a router to url mapper to the builder
   * @param routerToUrlMapper The url
   * @return {EndpointBuilder} The builder itself
   */
  withRouterToUrlMapper (routerToUrlMapper) {
    this._routerToUrlMapper = routerToUrlMapper
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder} The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride (responseOverride) {
    this._responseOverride = responseOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._routerToUrlMapper) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._routerToUrlMapper, this._mapper, this._refetchLogic, this._responseOverride)
  }
}

export default EndpointBuilder
