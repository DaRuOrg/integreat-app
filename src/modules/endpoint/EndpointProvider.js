import React from 'react'
import PropTypes from 'prop-types'
import Endpoint from './Endpoint'
import Error from '../common/containers/Error'

class EndpointProvider extends React.Component {
  static propTypes = {
    endpoints: PropTypes.arrayOf(PropTypes.instanceOf(Endpoint)).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    getEndpoint: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      getEndpoint: (endpointName) => {
        const endpoint = this.endpoints[endpointName]
        if (!endpoint) {
          throw new Error('The endpoint ' + endpointName + ' was not found. Maybe you didn\'t wrap the ' +
            'withFetcher(...) call in a <EndpointProvider /> or the endpoint is not registered')
        }
        return endpoint
      }
    }
  }

  constructor (props, context) {
    super(props, context)
    this.endpoints = props.endpoints.reduce((accumulator, endpoint) => {
      accumulator[endpoint.stateName] = endpoint
      return accumulator
    }, {})
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
