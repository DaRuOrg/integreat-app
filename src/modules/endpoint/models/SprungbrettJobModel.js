class SprungbrettModel {
  constructor ({id, title, city, url, isEmployment, isApprenticeship}) {
    this._id = id
    this._title = title
    this._city = city
    this._url = url
    this._isEmployment = isEmployment
    this._isApprenticeship = isApprenticeship
  }

  get id () {
    return this._id
  }

  get title () {
    return this._title
  }

  get city () {
    return this._city
  }

  get url () {
    return this._url
  }

  get isEmployment () {
    return this._isEmployment
  }

  get isApprenticeship () {
    return this._isApprenticeship
  }
}

export default SprungbrettModel
