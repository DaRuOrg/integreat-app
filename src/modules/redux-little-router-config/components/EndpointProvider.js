import React from 'react'
import PropTypes from 'prop-types'
import RouteConfig from '../RouteConfig'

class EndpointProvider extends React.Component {
  static propTypes = {
    routeConfig: PropTypes.instanceOf(RouteConfig).isRequired,
    children: PropTypes.element.isRequired
  }

  static childContextTypes = {
    matchRoute: PropTypes.func.isRequired
  }

  getChildContext () {
    return {
      matchRoute: (id) => this.props.routeConfig.matchRoute(id)
    }
  }

  render () {
    return React.Children.only(this.props.children)
  }
}

export default EndpointProvider
