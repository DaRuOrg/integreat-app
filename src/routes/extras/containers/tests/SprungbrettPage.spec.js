import { shallow, mount } from 'enzyme'
import React from 'react'
import ConnectedSprungbrettPage, { SprungbrettPage } from '../SprungbrettPage'
import SprungbrettJobModel from '../../../../modules/endpoint/models/SprungbrettJobModel'
import EndpointBuilder from '../../../../modules/endpoint/EndpointBuilder'
import createHistory from '../../../../modules/app/createHistory'
import createReduxStore from '../../../../modules/app/createReduxStore'
import EndpointProvider from '../../../../modules/endpoint/EndpointProvider'
import { Provider } from 'react-redux'
import LanguageModel from '../../../../modules/endpoint/models/LanguageModel'

describe('SprungbrettPage', () => {
  const location = 'augsburg'
  const language = 'de'

  const jobs = [
    new SprungbrettJobModel({
      id: '0', title: 'WebDeveloper', location: 'Augsburg', isEmployment: true, isApprenticeship: true
    }),
    new SprungbrettJobModel({
      id: '1', title: 'BackendDeveloper', location: 'Augsburg', isEmployment: true, isApprenticeship: false
    }),
    new SprungbrettJobModel({
      id: '2', title: 'Freelancer', location: 'Augsburg', isEmployment: false, isApprenticeship: true
    })
  ]

  const languages = [
    new LanguageModel('en', 'English'),
    new LanguageModel('de', 'Deutsch')
  ]

  it('it should render a Failure if type is not valid', () => {
    const component = shallow(
      <SprungbrettPage setLanguageChangeUrls={() => {}}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'bla'} />
    )
    expect(component).toMatchSnapshot()
  })

  it('it should render a selector and a list if type is valid', () => {
    const component = shallow(
      <SprungbrettPage setLanguageChangeUrls={() => {}}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'all'} />
    )
    expect(component).toMatchSnapshot()
  })

  it('it should return all jobs', () => {
    const sprungbrettPage = shallow(
      <SprungbrettPage setLanguageChangeUrls={() => {}}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'all'} />
    ).instance()
    expect(sprungbrettPage.getJobs()).toEqual(jobs)
  })

  it('it should return jobs in which an apprenticeship is possible', () => {
    const sprungbrettPage = shallow(
      <SprungbrettPage setLanguageChangeUrls={() => {}}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'apprenticeships'} />
    ).instance()
    expect(sprungbrettPage.getJobs()).toEqual([jobs[0], jobs[2]])
  })

  it('it should return jobs in which an employment is possible', () => {
    const sprungbrettPage = shallow(
      <SprungbrettPage setLanguageChangeUrls={() => {}}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'employments'} />
    ).instance()
    expect(sprungbrettPage.getJobs()).toEqual([jobs[0], jobs[1]])
  })

  it('should set language change urls', () => {
    const setLanguageChangeUrls = jest.fn()
    const sprunbrettPage = shallow(
      <SprungbrettPage setLanguageChangeUrls={setLanguageChangeUrls}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'employments'} />
    ).instance()
    expect(setLanguageChangeUrls.mock.calls).toHaveLength(1)
    expect(setLanguageChangeUrls).toHaveBeenCalledWith(sprunbrettPage.instance().mapLanguageToUrl, languages)
  })

  it('should update language change urls only on relevant prop change', () => {
    const setLanguageChangeUrls = jest.fn()
    const sprunbrettPage = shallow(
      <SprungbrettPage setLanguageChangeUrls={setLanguageChangeUrls}
                       languages={languages}
                       location={location}
                       language={language}
                       sprungbrett={jobs}
                       type={'employments'} />
    ).instance()
    expect(setLanguageChangeUrls.mock.calls).toHaveLength(1)
    sprunbrettPage.setProps({type: 'all', ...sprunbrettPage.props})
    expect(setLanguageChangeUrls.mock.calls).toHaveLength(2)

    sprunbrettPage.setProps({type: 'employments', ...sprunbrettPage.props})
    expect(setLanguageChangeUrls.mock.calls).toHaveLength(2)
  })

  describe('connect', () => {
    const sprungbrettEndpoint = new EndpointBuilder('sprungbrett')
      .withStateToUrlMapper(() => 'https://weird-endpoint/api.json')
      .withMapper(json => json)
      .withResponseOverride(jobs)
      .build()

    it('should map state to props', () => {
      const store = createReduxStore(createHistory, {
        router: {params: {location: location, language: language, type: 'apprenticeships'}},
        languageChangeUrls: {}
      })

      const sprungbrettPage = mount(
        <Provider store={store}>
          <EndpointProvider endpoints={[sprungbrettEndpoint]}>
            <ConnectedSprungbrettPage />
          </EndpointProvider>
        </Provider>
      ).find(SprungbrettPage)

      expect(sprungbrettPage.props()).toEqual({
        type: 'apprenticeships',
        location: location,
        language: language,
        sprungbrett: jobs,
        dispatch: expect.any(Function)
      })
    })
  })
})
