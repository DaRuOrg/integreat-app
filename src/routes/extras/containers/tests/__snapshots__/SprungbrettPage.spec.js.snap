// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprungbrettPage it should render a Failure if type is not valid 1`] = `
<Translate(Failure)
  error="This site does not exist"
/>
`;

exports[`SprungbrettPage it should render a selector and a list if type is valid 1`] = `
<div
  className="container"
>
  <SprungbrettSelector
    basePath="/augsburg/de/extras/sprungbrett"
    type="all"
  />
  <SprungbrettList
    jobs={
      Array [
        SprungbrettModel {
          "_id": "0",
          "_isApprenticeship": true,
          "_isEmployment": true,
          "_location": "Augsburg",
          "_title": "WebDeveloper",
          "_url": undefined,
        },
        SprungbrettModel {
          "_id": "1",
          "_isApprenticeship": false,
          "_isEmployment": true,
          "_location": "Augsburg",
          "_title": "BackendDeveloper",
          "_url": undefined,
        },
        SprungbrettModel {
          "_id": "2",
          "_isApprenticeship": true,
          "_isEmployment": false,
          "_location": "Augsburg",
          "_title": "Freelancer",
          "_url": undefined,
        },
      ]
    }
  />
</div>
`;
